domains/client/relayer/src/lib.rs

```diff
@@ -229,200 +229,132 @@ where
-        // we process that block instead while using the finalized block number to generate the MMR
-        // proof `finalized_block_number -1`
-        let block_number_to_process = match finalized_block.0.checked_sub(&One::one()) {
-            None => return Ok(()),
-            Some(number) => number,
+                .flatten()
+        {
+            // if this message should relay,
+            // check if the dst_chain inbox nonce is more than message nonce,
+            // if so, skip relaying since message is already executed on dst_chain
+            let relay_message = msg.nonce >= dst_channel_state.next_inbox_nonce;
+            if !relay_message {
+                log::debug!(
+                    "Skipping message relay from {:?} to {:?}",
+                    msg.src_chain_id,
+                    msg.dst_chain_id,
+                );
+            }
+            relay_message
+        } else {
+            should_relay
+        }
+    });
+
+    msgs.inbox_responses.retain(|msg| {
+        let id = (msg.channel_id, msg.nonce);
+        let should_relay = match api.should_relay_inbox_message_response(
+            best_hash,
+            msg.dst_chain_id,
+            id,
+        ) {
+            Ok(valid) => valid,
+            Err(err) => {
+                tracing::error!(
+                    target: LOG_TARGET,
+                    ?err,
+                    "Failed to fetch validity of inbox message response {id:?} for domain {0:?}",
+                    msg.dst_chain_id
+                );
+                false
+            }
         };
-        let block_hash_to_process = consensus_chain_client
-            .hash(block_number_to_process)?
-            .ok_or(Error::MissingBlockHeader)?;
-
-        let api = consensus_chain_client.runtime_api();
-        let block_messages: BlockMessagesWithStorageKey = api
-            .block_messages(block_hash_to_process)
-            .map_err(|_| Error::FetchAssignedMessages)?;
-
-        let filtered_messages = Self::filter_messages(
-            consensus_chain_client,
-            consensus_chain_client,
-            block_messages,
-        )?;
 
-        // short circuit if the there are no messages to relay
-        if filtered_messages.outbox.is_empty() && filtered_messages.inbox_responses.is_empty() {
-            return Ok(());
+        let channel_storage = ChannelStorage::new(msg.dst_chain_id);
+        if should_relay
+            && let Some(dst_channel_state) = channel_storage
+                .get_channel_state_for(&**consensus_client, msg.src_chain_id, msg.channel_id)
+                .ok()
+                .flatten()
+            && let Some(dst_chain_outbox_response_nonce) =
+                dst_channel_state.latest_response_received_message_nonce
+        {
+            // relay inbox response if the dst_chain did not execute is already
+            let relay_message = msg.nonce > dst_chain_outbox_response_nonce;
```

Comment:
Same question as above, but condition here is `>` instead of `>=` above, which is suspicious.

Reply:
They are checking different nonce of the dst chain, `next_inbox_nonce` and `latest_response_received_message_nonce`, they are updated in different way.

Reply:
Then this check should likely be `msg.nonce == dst_chain_outbox_response_nonce + 1` instead, right? Anything larger than that seems to be an indication of something not going according to plan.